File: 15.StrStr/next.cpp
class Solution {
public:
    vector<int> buildNext(string &s) {
        const int N = s.length();
        vector<int> next(N+1);

        int i = 0;
        int j = -1;
        next[0] = -1;

        while (i < N) {
            if (-1 == j || s[i] == s[j]) {
                i++;
                j++;
                next[i] = j;
            } else {
                j = next[j];
            }
        }

        // 最后一项就是整个字符串的前后缀的最长匹配
        return next;
    }
};

File: 15.StrStr/next.py
class Solution(object):
    def buildNext(self, s):
        N = 0 if not s else len(s)

        next = [0] * (N + 1)

        i = 0
        j = -1

        next[0] = -1

        while i < N:
            if -1 == j or s[i] == s[j]:
                i += 1
                j += 1
                next[i] = j
            else:
                j = next[j]
        return next
